src/meta/src/rpc/server.rs

```diff
@@ -171,316 +153,311 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
+mod tests {
+    use core::panic;
+    use std::net::{IpAddr, Ipv4Addr};
+
+    use risingwave_common::util::addr::HostAddr;
+    use risingwave_pb::common::WorkerType;
+    use risingwave_rpc_client::MetaClient;
+    use tokio::time::sleep;
+
+    use super::*;
+
+    static _SLEEP_SEC: u64 = 6;
+
+    /// Start `n` meta nodes on localhost. First node will be started at `meta_port`, 2nd node on
+    /// `meta_port + 1`, ...
+    async fn _setup_n_nodes(
+        n: u16,
+        meta_port: u16,
+        enable_recovery: bool, // TODO: remove. Is always false
+    ) -> Vec<(JoinHandle<()>, WatchSender<()>)> {
+        let meta_store = Arc::new(MemStore::default());
+
+        let mut node_controllers: Vec<(JoinHandle<()>, WatchSender<()>)> = vec![];
+        for i in 0..n {
+            // TODO: use http or https here?
+            let addr = format!("http://127.0.0.1:{}", meta_port + i);
+
+            let info = AddressInfo {
+                addr,
+                listen_addr: SocketAddr::new(
+                    IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
+                    meta_port + i,
+                ),
+                prometheus_addr: None, // TODO: use default here
+                dashboard_addr: None,
+                ui_path: None,
+            };
+            node_controllers.push(
+                rpc_serve_with_store(
+                    meta_store.clone(),
+                    info,
+                    Duration::from_secs(4),
+                    2,
+                    MetaOpts::test(enable_recovery),
+                )
+                .await
+                .unwrap_or_else(|e| panic!("Meta node{} failed in setup. Err: {}", i, e)),
+            );
+        }
+        sleep(Duration::from_secs(_SLEEP_SEC)).await;
+        node_controllers
     }
 
-    let catalog_manager = Arc::new(CatalogManager::new(env.clone()).await.unwrap());
-
-    let (barrier_scheduler, scheduled_barriers) =
-        BarrierScheduler::new_pair(hummock_manager.clone(), env.opts.checkpoint_frequency);
+    /// Check for `number_of_nodes` meta leader nodes, starting at `meta_port`, `meta_port + 1`, ...
+    /// Simulates `number_of_nodes` compute nodes, starting at `meta_port`, `meta_port + 1`, ...
+    ///
+    /// ## Returns
+    /// Number of nodes which currently are leaders. Number is not snapshoted. If there is a
+    /// leader failover in process, you may get an incorrect result
+    async fn _number_of_leaders(number_of_nodes: u16, meta_port: u16, host_port: u16) -> u16 {
+        let mut leader_count = 0;
+        for i in 0..number_of_nodes {
+            let local = "127.0.0.1".to_owned();
+            let port = meta_port + i;
+            let meta_addr = format!("http://{}:{}", local, port);
+            let host_addr = HostAddr {
+                host: local,
+                port: host_port + i,
+            };
+
+            let is_leader =
+                tokio::time::timeout(std::time::Duration::from_millis(100), async move {
```

Comment:
https://github.com/risingwavelabs/risingwave/pull/6466#discussion_r1054356572
https://github.com/risingwavelabs/risingwave/pull/6466#discussion_r1055144823

Reply:
Thank you very much for your guidance. I went for the approach we discussed and without the retry logic. A timeout is no longer necessary.